### DISCUSSION TITLE
Support automatically switching the host to the current guest resolution in full screen mode

### DISCUSSION THREAD

**Discussion by TheElf01** [86Box/86Box#4444] [Ideas]
Hi, sorry my english, hope i can explain my doubt

I have a question,  is possible to auto switch resolutions in fullscreen? i dont find any option to do...

Something like dosbox "resolution=original" option, 720x400, 640x400, 640x480, 640x200,  320x200, 720x350, etc etc

I have my windows desktop at 640x480 but VGA machines start at 720x400 for example, or 640x200 for CGA, etc etc  but 86Box always use my desktop resolution (640x480)

thanks

**Comment by OBattler** [comment]
I'm not sure what DOSBox does, but 86Box issues a resize whenever an emulated CRTC register is updated, so it often ends up in strange resolutions until the guest has finished programming all the CRTC registers. That makes me reluctant to implement such an "original resolution" feature as such programming would then set the host monitor to strange resolutions, which could have unintended, potentially even negative, effects.

**Comment by TheElf01** [comment]
Hi,  

> 
> 
> I'm not sure what DOSBox does, but 86Box issues a resize whenever an emulated CRTC register is updated, so it often ends up in strange resolutions until the guest has finished programming all the CRTC registers. That makes me reluctant to implement such an "original resolution" feature as such programming would then set the host monitor to strange resolutions, which could have unintended, potentially even negative, effects.

Hi,  thanks for reply

I dont know about 86Box internals (i did not check code) but at least, if you emulate a VGA machine, for sure need to change to 720x400 (for example for bios), 640x350 EGA,  CGA to 640x200 or 720×348 for hercules, etc etc  all standard resolutions

What other strange resolutions can be? you say that after change to, for example, 720x400 86box send weird random resolutions or something like this? im curious, thanks

But even if this happen,  a out of range error, and thats all no? no big deal

again thanks a lot for reply

**Comment by dhrdlicka** [comment]
This is pretty much dependent on the used display driver. 86Box, like any other Windows application, can't simply reprogram the graphic adapter on its own to change the resolution.

Modes like 640x480, 1024x768 are quite standard and there is no reason to for driver developers to leave them out, however, I am fairly sure no display driver will allow you to change to 640x350 or 640x200, not to mention some drivers seem to have dropped for lower color depths than 24 bits. Modern graphics adapters' EGA/CGA compatibility whatsoever tends to be pretty awful, with all sorts of glitches.

This is also useless with an LCD, since it would just stretch the image, which you can already do without changing the resolution. I guess it would be nice with a CRT, however, as I said, display drivers don't necessarily have to implement all 86Box emulated modes, so in the end we would have to resort to stretching anyway.

To reword what @OBattler said, reprogramming of the graphics adapter to change the resolution consists of multiple writes. Before the whole operation is finished, the adapter is pretty much in an undefined state. Considering that currently 86Box updates the resolution after each of these writes, it tends to set the resolution to very strange ones, until the guest finishes the operation.

**Comment by TheElf01** [comment]
> 
> 
> This is pretty much dependent on the used display driver. 86Box, like any other Windows application, can't simply reprogram the graphic adapter on its own to change the resolution.
> 
> Modes like 640x480, 1024x768 are quite standard and there is no reason to for driver developers to leave them out, however, I am fairly sure no display driver will allow you to change to 640x350 or 640x200, not to mention some drivers seem to have dropped for lower color depths than 24 bits. Modern graphics adapters' EGA/CGA compatibility whatsoever tends to be pretty awful, with all sorts of glitches.
> 
> This is also useless with an LCD, since it would just stretch the image, which you can already do without changing the resolution. I guess it would be nice with a CRT, however, as I said, display drivers don't necessarily have to implement all 86Box emulated modes, so in the end we would have to resort to stretching anyway.
> 
> To reword what @OBattler said, reprogramming of the graphics adapter to change the resolution consists of multiple writes. Before the whole operation is finished, the adapter is pretty much in an undefined state. Considering that currently 86Box updates the resolution after each of these writes, it tends to set the resolution to very strange ones, until the guest finishes the operation.


Custom resolutions are in the user side,  86Box dont need to do anything about this, just try to use the original resolution, like dosbox do for example (if not, just the closest one)

![IMG_20190709_173707](https://user-images.githubusercontent.com/16392474/60902660-49d14880-a270-11e9-83cc-e683d531ea81.jpg)

Can you see i create 720x350@50hz, and dosbox use this resolution, to show perfect Hercules Windows 3.1 desktop in my 31khz screen

![IMG_20190709_172633](https://user-images.githubusercontent.com/16392474/60902720-62416300-a270-11e9-9e7b-4b921e9176b0.jpg)

**Comment by OBattler** [comment]
Noone is talking about custom resolutions. We're talking about when the guest programs the emulated graphics card to a specific resolution. It's not a single command that instantly sets the emulated card to a particular resolution - the guest (be it the guest OS, the emulated machine's BIOS, or whatever) has to update 15+ CRTC registers, and until it's done, the CRTC is an undefined state. Now, every CRTC register change triggers a resize in 86Box, so in the interim state, the emulator could find itself in a strange interim resolution. Remember that vertical total and horizontal total are at least two CRTC registers each, so what prevents an interim resolution such as 0x350 while the CRTC registers are being programmed?

Edit: And unlike DOSBox, 86Box runs a real BIOS and a real period operating system, so the order in which the CRTC registers of the emulated graphics card get programmed is entirely unpredictable, so it's impossible to employ any kind of tricks, such as only updating resolution when the second of the two CRTC registers is written to.

**Comment by TheElf01** [comment]
> 
> 
> Noone is talking about custom resolutions. We're talking about when the guest programs the emulated graphics card to a specific resolution. It's not a single command that instantly sets the emulated card to a particular resolution - the guest (be it the guest OS, the emulated machine's BIOS, or whatever) has to update 15+ CRTC registers, and until it's done, the CRTC is an undefined state. Now, every CRTC register change triggers a resize in 86Box, so in the interim state, the emulator could find itself in a strange interim resolution. Remember that vertical total and horizontal total are at least two CRTC registers each, so what prevents an interim resolution such as 0x350 while the CRTC registers are being programmed?
> 
> Edit: And unlike DOSBox, 86Box runs a real BIOS and a real period operating system, so the order in which the CRTC registers of the emulated graphics card get programmed is entirely unpredictable, so it's impossible to employ any kind of tricks, such as only updating resolution when the second of the two CRTC registers is written to.

ok, thanks.  Sad there is no option, i was hopping to replace virtual machines with 86box

thanks again

**Comment by dhrdlicka** [comment]
I don't think the operating system would allow an application to voluntarily fry the monitor by setting it to some insane resolution. An option to automatically set the resolution to the closest one could be nice.

We will look into this for a future release.

**Comment by retrorepair** [comment]
Switchres is a viable option here.

In fact the late psakhis did exactly that: https://github.com/psakhis/86Box4crt

He says it's aimed at 15khz displays but works perfectly well with 31khz displays too

**Comment by innergrounds** [comment]
# Native host display mode-setting for fullscreen rendering (Linux OSD build)

This patch addresses [issue #395](https://github.com/86Box/86Box/issues/395):
*"Support automatically switching the host to the current guest resolution
in full screen mode."*

## Scope

**This patch only affects the Linux OSD build of 86Box** — the build
described in [`README-UNIX-MODE-WITH-OSD.txt`](https://github.com/86Box/86Box/blob/master/README-UNIX-MODE-WITH-OSD.txt),
where SDL2 talks directly to KMS/DRM via the `kmsdrm` video driver
(no Qt, no X, no Wayland).

The Qt builds (Linux/Mac/Windows) are not modified. The cfg keys are read
unconditionally, but the patch logic lives entirely in `src/unix/unix.c`
and `src/unix/unix_sdl.c`, so other builds ignore the keys.

The default behavior (without setting any cfg keys) is byte-identical to
upstream master, including for OSD builds.

## What it does

When the emulated machine changes its video mode (e.g., DOS BIOS POST at
720×400, then Mode 13h at 320×200, then Win95 at 1024×768), the host
display performs a real DRM mode-set to match. No more scaling the
emulated framebuffer into a fixed-size SDL window — the host adopts the
emulated dimensions directly.

Two modes:

- `video_modeset = 1` (exact): `SDL_SetWindowDisplayMode` with the exact
  emulated dimensions, then toggle `SDL_WINDOW_FULLSCREEN` to apply.
  Targets CRTs and any display that accepts arbitrary modes.
- `video_modeset = 2` (centered): use `SDL_GetClosestDisplayMode` to find
  an EDID-listed mode, switch to that, render the source 1:1 centered
  with black borders. LCD-friendly fallback.

## What changed

Five files, ~70 lines added, 3 removed:

```
 src/86box.c               |  4 ++++
 src/config.c              |  4 ++++
 src/include/86box/86box.h |  9 +++++++++
 src/unix/unix.c           |  4 ++--
 src/unix/unix_sdl.c       | 53 ++++++++++++++++++++++++++++++++++++++++++++++++++++-
```

### `src/unix/unix.c` (2 lines)

In the main loop's screen-resize block:
- Removed the `&& !video_fullscreen` guard. Upstream silently dropped
  resize signals from the emulated card when in fullscreen mode; this is
  what makes the SDL window stay at native resolution and forces
  texture-scaling.
- Changed the post-handle store of `doresize_monitors[0]` from `1` to `0`
  to clear the flag. The original `1` triggers an infinite resize loop
  once the guard above is removed.

These are gated on `video_modeset != 0`. With modeset off, behavior is
identical to upstream.

### `src/unix/unix_sdl.c` (~50 lines added, 1 removed)

- New macro `modeset_log(...)` wrapping `fprintf(stderr, ...)` behind
  `video_modeset_verbose`.
- In `sdl_real_blit()`, fullscreen branch: when `video_modeset ==
  VIDEO_MODESET_CENTERED`, render 1:1 centered. The existing
  `SDL_RenderClear` already wipes borders to black.
- In `plat_resize()`:
  - Out-of-range bounds check (configurable via cfg) at function entry.
    Some emulated cards transiently emit absurd dimensions like `720x2001`
    during mode transitions; this is a defensive filter.
  - `VIDEO_MODESET_EXACT` branch: pass the emulated dimensions verbatim
    to `SDL_SetWindowDisplayMode`, then toggle fullscreen to apply.
  - `VIDEO_MODESET_CENTERED` branch: call `SDL_GetClosestDisplayMode`
    instead, accept whatever the EDID supports, let `sdl_real_blit`
    center the source.

Both branches gate on a `last_fs_w/last_fs_h` cache so mode-set fires only
on actual dimension changes, not every frame.

### Cfg keys (`src/86box.c`, `src/include/86box/86box.h`, `src/config.c`)

Four new keys in `[General]`, all default to 0:

| Key                       | Type | Meaning                                      |
|---------------------------|------|----------------------------------------------|
| `video_modeset`           | int  | 0 = off (default), 1 = exact, 2 = centered   |
| `video_modeset_max_w`     | int  | 0 = no limit; otherwise width upper bound    |
| `video_modeset_max_h`     | int  | 0 = no limit; otherwise height upper bound   |
| `video_modeset_verbose`   | int  | 0 = quiet (default), 1 = log to stderr       |

Plus three `#define`s in the header:

```c
#define VIDEO_MODESET_OFF      0
#define VIDEO_MODESET_EXACT    1
#define VIDEO_MODESET_CENTERED 2
```

## Behavior matrix

| Display | `video_modeset = 1` (exact)               | `video_modeset = 2` (centered)        |
|---------|-------------------------------------------|---------------------------------------|
| CRT     | Switches to exact emulated mode 1:1.      | Closest mode + 1:1 centered render.   |
|         | Pixel-perfect, period-authentic.          | Same effect as 1 in practice.         |
| LCD     | Standard emulated modes (640×480, 800×600,| Always switches to a EDID-listed mode,|
|         | 1024×768, etc.) work fine. Non-standard   | source rendered 1:1 centered with     |
|         | modes (320×200, 720×400) typically refused| black bars. Works for all modes.      |
|         | by the LCD — no signal. Use mode 2.       |                                       |

Upstream behavior (`video_modeset = 0`) is unchanged.

## Diagnostic output

With `video_modeset_verbose = 1`, mode changes log to stderr:

```
86box: plat_resize fs(2) 320x400 -> closest 640x480@60
86box: plat_resize fs(2) 720x400 -> closest 720x480@60
86box: plat_resize fs(2) 1024x768 -> closest 1024x768@60
86box: plat_resize ignoring out-of-range 720x2001 (limit 1600x1200)
```

The `out-of-range` line is unconditional even with verbose=0, since it
indicates a defensive filter activation worth surfacing.

## Caveats / open questions

Things I'd like reviewer guidance on before opening a PR:

1. **Linux OSD build only.** This is a deliberate scope decision — OSD
   builds are exactly where users are most likely to want native
   mode-setting (appliance/dedicated retro-PC use, often paired with a
   real CRT). Qt builds typically run inside a desktop session where
   userspace mode-setting fights the compositor. If you want this to
   apply to Qt builds on Linux/Mac/Windows too, point me at the right
   places in those renderers.
2. **No GUI integration.** Cfg-file-only. Adding a Display settings
   panel dropdown in the Qt UI is straightforward but enlarges the patch.
   Want it in the same PR or as a follow-up?
3. **Hardcoded monitor index 0.** The mode-set targets monitor 0 only.
   Multi-monitor 86Box setups would need to loop over `doresize_monitors[]`.
   Not implemented because I have no multi-monitor testbed.
4. **Refresh rate not threaded.** Currently `want.refresh_rate = 0` lets
   SDL pick. For period-authentic CRT use, DOS games run at 70 Hz and
   Win9x at 60 Hz. Could pull the actual rate from the emulated card's
   video timing. Significant additional work; not done.
5. **Macro vs. inline function** for `modeset_log`. I went with a macro
   for brevity and lazy arg evaluation. If 86Box has a preferred logging
   idiom, happy to adopt it.

## Testing performed

- Stock-default (no cfg keys): identical behavior to upstream master,
  verified by visual A/B on the OSD build.
- `video_modeset = 1`: tested with DOS (320×200, 720×400 text), Win 3.1
  (640×480), Win95 boot through to desktop (640×400 → 1024×768), with
  both the bound check and `last_fs_w/h` cache active.
- `video_modeset = 2`: same scenarios; centered display with sharp pixels.
- Bounds check verified by setting `video_modeset_max_w = 500` and
  observing rejection of normal dimensions.
- Verbose=0 produces zero log output during normal operation (other than
  the unconditional out-of-range line if it triggers).

Tested on Debian 13 / SDL2 2.32.4 / Intel HD 4600 / kmsdrm driver,
running on tty1 with no X session.

## Build instructions

Standard OSD build per [`README-UNIX-MODE-WITH-OSD.txt`](https://github.com/86Box/86Box/blob/master/README-UNIX-MODE-WITH-OSD.txt):

    cmake -S . -B build/regular --preset regular -D QT=OFF -D PREFER_STATIC=ON
    cmake --build build/regular -j 4

Apply patch:

    git apply video_modeset.patch
    # or, if line numbers have drifted:
    patch -p1 < video_modeset.patch

Patch attached.
[video_modeset.patch](https://github.com/user-attachments/files/27455748/video_modeset.patch)

**Reply by druid-426** [reply]
Thanks for the patch. I have to say though I've been having a tough time getting it to operate as it should on my R5 340x. It seems to be handling resolution changes but I'm locked to whatever resolutions are read from EDID. I can't get it to send arbitrary signals. I've tried a whole bunch of things like forcing legacy radeon driver and updating my kernel to 6.19 for better amdgpu support. I looked into it a little and I think the amdgpu driver is just more restrictive than i915.

Anyway, if you have suggestions on hardware I'd love to hear it. As I don't have an intel platform, I'm considering swapping out the amd card with something like a GT 710. 

Thanks again!

**Comment by innergrounds** [comment]
Hey,
There are 3 modes, one is disable, one queries the closest resolution and
one just renders without checking (this is what I use on a real CRT).
I use an old  Lenovo M93p Tiny which has an actual real VGA output (not a
DP to VGA port).

**Reply by druid-426** [reply]
I see. My card has analog out through it's DVI-I port. I am using an adapter to vga but it should be handled the same as a regular vga port as far as I know and it's completely passive. I'm curious though, Is your monitor able to handle low frequency resolutions like 320x200 natively or is your output being line doubled?

**Reply by OBattler** [reply]
And here we run into another problem - our 320x200 on EGA and (S)VGA is always line doubled, but that's also true for the real hardware, unlike for eg. the CGA.

**Reply by mgtroyas** [reply]
DOSBox does exactly the same. If you disable Scale2X to always render original resolution, the exception are those modes, they're sent line doubled to the output.

**Reply by mgtroyas** [reply]
> I see. My card has analog out through it's DVI-I port. I am using an adapter to vga but it should be handled the same as a regular vga port as far as I know and it's completely passive. I'm curious though, Is your monitor able to handle low frequency resolutions like 320x200 natively or is your output being line doubled?

I guess it's not the monitor but the VGA card which line doubles, in the monitor OSD you can see it detects 720x400 resolution (720 because that's the resolution of text mode and the monitor cannot guess horizontal resolution from the analogue signal).

**Comment by innergrounds** [comment]
No, a DP to VGA is not passive and you will never get the VGA native
resolution.
I tested the exact same case you have but that depends on the converter and
basically does not work at low resolution.
With the old machine I use, resolution switches perfectly on the CRT.

**Comment by mgtroyas** [comment]
Hello, seeing how amazingly 86box runs on fullscreen on a real CRT with zero lag, I hope I can help on any regard. My "modern" Windows 11 PC has a Radeon RX 580 connected to a 19" SyncMaster using a StarTech DP2VGAHD adapter. I've created many custom resolutions for use with RetroBat and DOSBox, and could adapt them to also test anything on a Windows 86Box test build.

I also have a Windows XP SFF build with a Radeon 7750 and VGA output, directly connected to the CRT for any other testing needed. I have there installed hundreds of Windows games, and a huge DOSBox collection that uses the 640x400@70Hz video mode, not available on Windows 7 and newer.

Right now what I do is manually switch the CRT video mode with a third party tool whenever the 86box VM changes it's emulated video mode, but 86Box detecting the change on screen size and refresh rate and sending it to the real monitor on the fly would be a dream. Perhaps to avoid false positives, a delay of 1 second or so to avoid quick momentary changes could avoid weird behaviours.

**Comment by nbouayad-wincomplm** [comment]
Sorry to say that but as far as I know (and tested), there is no 1 to 1 DP/HDMI or so converter to VGA.
Also, Windows is not able to drive the setup you want (because of rendering abstraction layers such as DirectX or so).
The OSD build.with my patch on a old machine with a real VGA output which can run a minimal Debian works just fine (I use 13 I think).
I have an automatically starting VM with Windows 95 RTM where I can switch to dos and run all the classics without any problem.
That Lenovo cost was 150 dollars max.
If you want information on how to setup such dedicated appliance I can share the information.
It might not be technically perfect but it works for me.


Sent from a mobile device.
________________________________

**Reply by mgtroyas** [reply]
In my experience, the DP/HDMI converter outputs a resolution and refresh rate equal to the one requested via VGA input, so inside that range the converter is transparent for the application. This range is limited by the upper and lower bandwidth (MHz) it allows by design:
- On the lower limit, it's enough for any 30kHz CRT monitor, I've managed 640x400@70Hz without problems, in fact some monitors don't even allow so low horizontal frequency. I don't have a lower refresh CRT like 15kHz to test.
- On the upper limit, it depends on the converter quality: the cheap ones, I guess equivalent to DVI Single-Link, are enough for typical 70KHz monitors, but for 96kHz+ monitors you need a higher quality converter like the StarTech DP2VGAHD. I've managed 2048x1536@60Hz without problem on mine.
- Out of that compatible range, you just get no output, no damage is done to the monitor.

So i guess 86Box only has to send the OS the order to switch video mode, on the right monitor (I use dual monitor setup) indifferent of being connected through VGA or an adapter to the CRT.

I've also built for fun my Windows XP build, and have a laptop for Windows 98, but unfortunately time passes by and that hardware will eventually fail, that's the value of projects like 86Box and DOSBox. Also having many systems emulated on a single modern one brings space and money advantages, XP era hardware is cheap as of today, but we know what happened to older hardware prices recently and that'll also happen as the rolling window of retro systems advances. The day the last CRT has also failed we will have to conform with OLED and CRT shaders, which is not bad at all, and finally we will also fail and nobody will care about the difference because they didn't experience the original, but for now let's take advantage of the CRTs while they and us still work.

**Comment by innergrounds** [comment]
I added the centered mode for the resolution issues.
Could not get exactly the DOS resolution with the adapter (might be due to
the screen maybe).

**Comment by druid-426** [comment]
Update to my Odyssey in getting this set up. Im starting to realize I'm not too good with computers :^}. Anyway it turns out that 720x400 is a standard VESA displaymode. For some reason my monitor's edid had it disabled! I gave it a quick toggle with wxEDID and like magic, scaling is perfect. In fact, so perfect it's not actually scaling anything! THIS IS NATIVE! 
I've exhausted all tricks I could think of for an arbitrary-ly syncable crt besides just adding a bunch of modes to my edid (which I don't feel like doing) or changing my "Video="  parameters in kernel command like every time I run into a funky resolution. (would require a host reboot each time).
I'm hoping the $10 gt 730  I've got in the mail along side this "metamodes" thing that according to Google only exists in the proprietary nvidia drivers will somehow fix all my problems.
Thanks for reading this long post if you did. I'll give updates once I have the new hardware installed!
